### PR TITLE
Ability to provide extra fields

### PR DESCRIPTION
**Pull request by naps62**
I'm trying to incorporate this into a current project. I'm having a problem though. I want to provide atomic transitions where additional fields are required.

E.g: when transitioning from state A to state B, field "extra" is needed on the struct, but I want to update both fields (`:state` and `:extra`) in a single query.

This PR is based off of the work @TalkingQuickly did on #54 , but goes a bit further. I added an `extra` field (defaulting to `%{}`) to all operations, and that allows me to do this:

```elixir
def guard_transition(struct, state, extra) do
  changeset = Struct.changeset(extra) |> put_change(:state, state)

  if is_nil(Ecto.Changeset.get_field(changeset, :some_required_field)) do
    {:error, "some_required_field is mandatory when transitioning to #{state}"}
  else
    true
end

def persist(struct, state, extra) do
  {:ok, updated_struct} = 
    Struct.changeset(extra)
    |> put_change(:state, state)
    |> Repo.update()
    
    updated_struct
end
```

However, this is a slight breaking change, since all the overridable functions now receive an extra argument. Even though it can be ignored, that requires updating existing state machines to include it

However it feels like a safe enough change that provides abilities not possible with the current API

PS: I haven't updated the readme or the documentation yet. I'll wait for feedback on this change before proceeding with that

### CODE REVIEW

File: test/support/test_state_machine_with_guard.exs [hunk 1]
```diff
@@ -7,10 +7,11 @@ defmodule MachineryTest.TestStateMachineWithGuard do
       "partial" => "completed"
     }
 
-  def guard_transition(struct, "completed") do
+  def guard_transition(struct, "completed", _extra) do
     # Code to simulate and force an exception inside a
     # guard function.
     if Map.get(struct, :force_exception) do
+      IO.inspect "raising"
```
Comment (joaomdmoura) [inline test/support/test_state_machine_with_guard.exs]: Missing inspect 😄

**Review by joaomdmoura — COMMENTED** [review]
LGTM, I like the idea, don't think we need to care about the breaking change, as you mentioned this add new functionality, and we can cut a new breaking version using SEMVER.

File: test/support/test_state_machine_with_guard.exs (hunk 1 above)
Comment (naps62) [reply test/support/test_state_machine_with_guard.exs]: whoops! sorry :smile:

**Comment by joaomdmoura** [comment]
Fixed conflicts, will review and merge later today, cut a new version later in the week probably since there were other PRs merged as well

**Comment by Nuno-Silva-Bose** [comment]
@joaomdmoura do you still intend to merge this PR?

Was checking the PR and `extra` is mandatory. Shouldn't it be put as optional? There are several scenarios where extra parameters are not needed.

Thanks!

**Comment by RodolfoSilva** [comment]
@joaomdmoura is there any chance of this PR be merged?

**Comment by kyleboe** [comment]
@joaomdmoura are you planning on still maintaining this library? If not, I'd be happy to take over maintenance.

**Comment by joaomdmoura** [comment]
I'm picking up the work on this one oh #85 , I decided to go with another implementation that wont brake any implementation for current projects using this that decide to update.

Because of that I'm closing this in favor of the other one.
This was a great contribution tho, as it was the main inspiration for the work on #85.

**Pull request closed without merging**